lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,374 @@
+
+import numpy as np
+from iris.coords import AuxCoord, DimCoord
+from iris import Constraint
+from iris.cube import Cube
+from numpy.linalg import lstsq
+from improver.spotdata.common_functions import (nearest_n_neighbours,
+                                                node_edge_test,
+                                                isclose)
+from improver.spotdata.read_input import data_from_dictionary
+from improver.constants import (R_DRY_AIR,
+                                CP_DRY_AIR)
+
+
+class ExtractData(object):
+    """
+    A series of methods for extracting data from grid points to derive
+    diagnostic values at off grid positions.
+
+    """
+
+    def __init__(self, method='use_nearest'):
+        """
+        The class is called with the desired method to be used in extracting/
+        interpolating data to the site of interest from gridded data.
+
+        """
+
+        self.method = method
+
+    def process(self, cube, sites, neighbours, ancillary_data, additional_data,
+                no_neighbours=9):
+        """
+        Call the correct function to enact the method of data extraction
+        specified. This function also handles multiple timesteps, consolidating
+        the resulting data cubes into an Iris.CubeList.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            Cube of diagnostic data from which to extract spotdata.
+
+        sites : dict
+            A dictionary containing the properties of spotdata sites.
+
+        neighbours : numpy.array
+            Array of neigbouring grid points that are associated with sites
+            in the SortedDictionary of sites.
+
+        additional_data : dict
+            A dictionary containing any supplmentary time varying diagnostics
+            that are needed for the selected extraction method.
+
+        ancillary_data : dict
+            A dictionary containing additional model data that is needed.
+            e.g. {'orography': <cube of orography>}
+
+        no_neighbours : int
+            Number of grid points about the site to consider when calculating a
+            local temperature lapse rate using orography variation.
+            e.g. consider a 5x5 grid of points -> no_neighbours = 25.
+
+        Returns:
+        --------
+        cube : iris.cube.Cube
+            An irregularly (i.e. non-gridded) cube of diagnostic data extracted
+            at the spotdata sites.
+
+        """
+        if self.method == 'use_nearest':
+            return self.use_nearest(cube, sites, neighbours)
+        elif self.method == 'orography_derived_temperature_lapse_rate':
+            orography = data_from_dictionary(ancillary_data, 'orography').data
+            return self.orography_derived_temperature_lapse_rate(
+                cube, sites, neighbours, orography,
+                no_neighbours=no_neighbours)
+        elif self.method == 'model_level_temperature_lapse_rate':
+            height_50 = Constraint(height=50)
```

Comment:
Added ability to set these using the hybrid height level indices. 1 and 3 (10m and 30m) are the values currently used in bestdata2, so I have set these as the defaults.